exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/TraceMarshaler.java

```diff
@@ -524,21 +524,14 @@ private static int computeSize(
 
   private static Map<Resource, Map<InstrumentationLibraryInfo, List<SpanMarshaler>>>
       groupByResourceAndLibrary(Collection<SpanData> spanDataList) {
-    Map<Resource, Map<InstrumentationLibraryInfo, List<SpanMarshaler>>> result = new HashMap<>();
+    IdentityHashMap<Resource, Map<InstrumentationLibraryInfo, List<SpanMarshaler>>> result =
+        new IdentityHashMap<>();
     for (SpanData spanData : spanDataList) {
-      Resource resource = spanData.getResource();
       Map<InstrumentationLibraryInfo, List<SpanMarshaler>> libraryInfoListMap =
-          result.get(spanData.getResource());
-      if (libraryInfoListMap == null) {
-        libraryInfoListMap = new HashMap<>();
-        result.put(resource, libraryInfoListMap);
-      }
+          result.computeIfAbsent(spanData.getResource(), unused -> new IdentityHashMap<>());
```

Comment:
could this be where the additional allocations are coming from? Something to do with the lambda invocation?

Reply:
This codepath isn't hit by that benchmark, I copied in the code for grouping to this "direct proto marshal" code which had diverged.